pysheaf.py

```diff
@@ -52,11 +52,49 @@ def isCoface(self,index,orientation=None):
             return (index,orientation) in [(cf.index,cf.orientation) for cf in self.cofaces]
         
 class CellComplex:
-    def __init__(self,cells=[]):
+    def __init__(self,cells=None):
         """Construct a cell complex from its cells"""
-        self.cells=cells
+        if cells == None:
+            self.cells=[]
+        else:
+            self.cells=cells
+
+        # Register cell names with indices into the self.cells list        
+        self.cell_dict={}
+        self.coface_dict={}
+        for i,c in enumerate(self.cells):
+            if c.name == None: # Build names if not present
+                c.name = str(i)
+            self.cell_dict[c.name]=i
+
+    def add_cell(self,cell):
+        """Add a cell to the cell complex"""
+
+        if cell.name == None: # Construct a name if needed
+            cell.name = str(len(self.cells))
+
+        self.cell_dict[cell.name]=len(self.cells)
+        self.cells.append(cell)
+
+    def add_cells_from(self,cells):
+        for c in cells:
+            self.add_cell(c)
+
+    def add_coface(self,names,coface):
+        """Add a coface to the cell complex, referenced by a pair of cell names.  The names argument is assumed to be a pair: (face,coface).  The .index attribute for the coface is ignored.  If the cells aren't present, this will raise KeyError."""
+        # Look up which cells are involved...
+        source=self.cell_dict[names[0]]
+        cf=copy.deepcopy(coface)
```

Comment:
Line 87 - I believe you construct cofaces here, not call existing ones. I am thinking:
```
### names = (face.id, coface.id) 
def add_coface(self, names, orientation):
    coface = Coface(index=names, orientation)
    self.cofaces[names] = coface
```